### PR TITLE
ddl: fix a bug when format generation expressions.

### LINKED ISSUE #5253: Create JSON generated field table,  The integrity of creating a table is broken
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
Create a new table contains json generated column

CREATE TABLE `xxxxxxxx` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `x1` bigint(20) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x1')) STORED NOT NULL ,
  `x2` bigint(20) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x2')) VIRTUAL ,
  `x3` bigint(20) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x3')) VIRTUAL,
  `x4` json GENERATED ALWAYS AS (json_unquote(json_extract(`json_field`,'$.x4'))) VIRTUAL ,
  `x5` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x5')) VIRTUAL ,
  `x6` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x6')) VIRTUAL ,
  `x7` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x7')) VIRTUAL ,
  `x8` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x8')) VIRTUAL ,
  `x9` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x9')) VIRTUAL ,
  `x10` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x10')) VIRTUAL ,
  `x11` varchar(50) GENERATED ALWAYS AS (json_unquote(json_extract(`json_field`,'$.x11'))) VIRTUAL ,
  `x12` datetime GENERATED ALWAYS AS (cast(json_unquote(json_extract(`json_field`,'$.x12')) as datetime)) VIRTUAL ,
  `x13` datetime GENERATED ALWAYS AS (cast(json_unquote(json_extract(`json_field`,'$.x13')) as datetime)) VIRTUAL ,
  `json_field` json NOT NULL ,
  `x15` datetime GENERATED ALWAYS AS (cast(json_unquote(json_extract(`json_field`,'$.x15')) as datetime)) VIRTUAL ,
  `x16` int(11) GENERATED ALWAYS AS (json_extract(`json_field`,'$.x16')) VIRTUAL,
  `x17` tinyint(1) NOT NULL DEFAULT '1' ,
  PRIMARY KEY (`id`),
  UNIQUE KEY `idx_x1` (`x1`)
) ENGINE=InnoDB AUTO_INCREMENT=1388 DEFAULT CHARSET=utf8 ;
 

2. What did you expect to see?

3. What did you see instead?

Error Code: 1105. line 1 column 70 near ")" (total length 71)	20.125 sec

When Execute the DDL again：the server response the table already exists。

Besides the problem above which is acceptable, the another problem could be devastating **the tidb-server cannot restart after stopped。** 
**It seems that the integrity of creating a table is broken。**
log：2017/11/28 19:10:05.761 domain.go:486: [error] [ddl] new domain failed line 1 column 80 near ")" (total length 81)

4. What version of TiDB are you using (`tidb-server -V`)?

The newest version on docker.  the server is broken,so i cannot run  (`tidb-server -V`)

### PR DESCRIPTION
**Pull request by hicqu**
fix #5253 
When we store generation expressions into `ColumnInfo`, we should remove redundant blanks in order to print it more pretty. There is a bug when we do this, which is we remove all blanks instead of redundant blanks. This PR fix this.

### CODE REVIEW

File: util/stringutil/string_util.go [hunk 1]
```diff
@@ -236,16 +236,25 @@ func DoMatch(str string, patChars, patTypes []byte) bool {
 	return sIdx == len(str)
 }
 
-// RemoveBlanks removes all blanks, returns a new string.
-func RemoveBlanks(s string) string {
+// RemoveRedundantBlanks removes all redundant blanks, returns a new string.
+func RemoveRedundantBlanks(s string) string {
```
Comment (shenli) [inline util/stringutil/string_util.go]: Could you add an unit test for this?

File: util/stringutil/string_util.go [hunk 2]
```diff
@@ -236,16 +236,25 @@ func DoMatch(str string, patChars, patTypes []byte) bool {
...
 	var buf = new(bytes.Buffer)
 	var cbuf [6]byte
 	for _, c := range s {
-		if c == rune(' ') || c == rune('\t') || c == rune('\r') || c == rune('\n') {
-			continue
+		if c == rune('\t') || c == rune('\r') || c == rune('\n') {
```
Comment (coocood) [inline util/stringutil/string_util.go]: How about using a variable to store if the last rune is blank, then continue?

File: util/stringutil/string_util.go (hunk 1 above)
Comment (hicqu) [reply util/stringutil/string_util.go]: There are unittests in `util/stringutil/string_util_test.go`.

**Comment by hicqu** [comment]
@zimulala PTAL.

**Comment by hicqu** [comment]
@coocood PTAL thanks.

**Comment by coocood** [comment]
LGTM

**Comment by neil4dong** [comment]
Please refer to this issue #5253 ， I believe the test is not enough。

**Comment by XuHuaiyu** [comment]
It would be a mistake if the expression is length("a `n_blanks` b")

**Comment by hicqu** [comment]
Yes, the final solution is serialize AST back to string. Maybe now it's time to add this interface.

**Comment by hicqu** [comment]
@XuHuaiyu @coocood PTAL thanks. Now I make  a quick fix for using string literal in generated expressions. But If we want to format generation expressions perfact, we need do it on AST directly. Maybe do it later is better.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by coocood** [comment]
@hicqu
The escape seems not handled properlly. 
Do we have to replace blanks?
How about just keep the original, and remove the blanks when ast to string is done?

**Comment by hicqu** [comment]
OK. It seems we can't avoid writing AST back to string anymore.

**Comment by hicqu** [comment]
Now we can remove redundant blanks correctly with writing AST to string.

**Comment by hicqu** [comment]
@XuHuaiyu @coocood PTAL thanks.

**Comment by hicqu** [comment]
ping @XuHuaiyu @coocood

**Comment by coocood** [comment]
@hicqu CI failed

**Comment by hicqu** [comment]
ping @XuHuaiyu @coocood

**Comment by coocood** [comment]
@hicqu 
The `ValueExpr` use its `Text` to format, but not all `ValueExpr`s set `Text` in parser.

We should either set text for all value expressions or change the implementation of Format.

**Comment by hicqu** [comment]
@coocood address comments. PTAL again, thanks.

**Comment by shenli** [comment]
@coocood @XuHuaiyu PTAL

**Comment by coocood** [comment]
@hicqu 
Some types of the `ValueExpr` (Time for example) should be quoted when it converts to string.
But on the other hand, those types will not ever found in `ValueExpr`.

So we should either remove those types or add quotes to them.

**Comment by hicqu** [comment]
Now we don't support `time/timestamp` literal, so I prefer remove them.

**Comment by hicqu** [comment]
@coocood Now we support `TIME` and `TIMESTAMP` literals, and I add test cases to show that. And I remove `enum`, `set` and `JSON`.

File: ast/format_test.go [hunk 3]
```diff
@@ -24,13 +24,19 @@ func (ts *testAstFormatSuite) TestAstFormat(c *C) {
...
+		{`true`, `TRUE`},
 		{`350`, `350`},
 		{`345.678`, `345.678`},
 		{`1e-12`, `0.000000000001`},
 		{`null`, `NULL`},
 		{`"Hello, world"`, `"Hello, world"`},
 		{`'Hello, world'`, `"Hello, world"`},
 		{`'Hello, "world"'`, `"Hello, \"world\""`},
+		{`x'bcde'`, "\xbc\xde"},
+		{`b'111111'`, `?`},
```
Comment (coocood) [inline ast/format_test.go]: Why the out put is `?`

File: ast/format_test.go [hunk 4]
```diff
@@ -24,13 +24,19 @@ func (ts *testAstFormatSuite) TestAstFormat(c *C) {
...
+		{`time'10:10:10.123'`, `timeliteral("10:10:10.123")`},
```
Comment (coocood) [inline ast/format_test.go]: The output is different.

Comment (coocood) [inline ast/format_test.go]: This is parsed as a function, so there is no Literal of time type.

File: ast/format_test.go (hunk 3 above)
Comment (hicqu) [reply ast/format_test.go]: 111111 is  63. `?` is the char of `63`.

File: ast/format_test.go [hunk 5]
```diff
@@ -24,13 +24,21 @@ func (ts *testAstFormatSuite) TestAstFormat(c *C) {
...
 		{`350`, `350`},
 		{`345.678`, `345.678`},
-		{`1e-12`, `1e-12`},
+		{`1e-12`, `0.000000000001`},
 		{`null`, `NULL`},
 		{`"Hello, world"`, `"Hello, world"`},
 		{`'Hello, world'`, `"Hello, world"`},
 		{`'Hello, "world"'`, `"Hello, \"world\""`},
+		{`_utf8'你好'`, `"你好"`},
+		{`x'bcde'`, "\xbc\xde"},
```
Comment (coocood) [inline ast/format_test.go]: I think we should make the output a binary literal, the output doesn't look right.

File: ast/format_test.go [hunk 6]
```diff
@@ -24,13 +24,21 @@ func (ts *testAstFormatSuite) TestAstFormat(c *C) {
...
+		{`b'111111'`, `?`},
```
Comment (coocood) [inline ast/format_test.go]: We should output binary literal string instead of '?'.

**Comment by hicqu** [comment]
@coocood PTAL thanks.

File: ast/expressions.go [hunk 7]
```diff
@@ -87,7 +88,22 @@ func (n *ValueExpr) Format(w io.Writer) {
...
+		// Hex literal has unsigned flag but bit literal not.
+		// See `types.DefaultTypeForValue`
+		if n.Type.Flag&mysql.UnsignedFlag != 0 {
+			sb = bytes.NewBufferString("x'")
+			for _, b := range n.GetBinaryLiteral() {
+				fmt.Fprintf(sb, "%s", strconv.FormatInt(int64(b), 16))
+			}
+		} else {
+			sb = bytes.NewBufferString("b'")
+			for _, b := range n.GetBinaryLiteral() {
```
Comment (coocood) [inline ast/expressions.go]: Use `BinaryLiteral.ToBitLiteralString`.

File: ast/expressions.go [hunk 8]
```diff
@@ -87,7 +88,22 @@ func (n *ValueExpr) Format(w io.Writer) {
 	case types.KindMysqlDecimal:
 		s = n.GetMysqlDecimal().String()
 	case types.KindBinaryLiteral:
-		s = n.GetBinaryLiteral().ToString()
+		var sb *bytes.Buffer
+		// Hex literal has unsigned flag but bit literal not.
+		// See `types.DefaultTypeForValue`
+		if n.Type.Flag&mysql.UnsignedFlag != 0 {
+			sb = bytes.NewBufferString("x'")
```
Comment (coocood) [inline ast/expressions.go]: Use `BinaryLiteral.String()`.

Comment (hicqu) [reply ast/expressions.go]: `String` doesn't suite here. `select "0xbcde"` has different field type from `select x'bcde'`. you can see it if you try it in MySQL cli with `--column-type-info`.

Comment (coocood) [reply ast/expressions.go]: OK, then this will do.
```
s = fmt.Sprintf("x'%x'", n.GetBytes())
```

File: types/binary_literal.go [hunk 9]
```diff
@@ -99,6 +99,17 @@ func (b BinaryLiteral) ToBitLiteralString(trimLeadingZero bool) string {
 	return fmt.Sprintf("b'%s'", string(ret))
 }
 
+func (b BinaryLiteral) ToHexLiteralString() string {
+	if len(b) == 0 {
```
Comment (coocood) [inline types/binary_literal.go]:
```
return fmt.Sprintf("x'%x'", []byte(b))
```

**Comment by hicqu** [comment]
@coocood @XuHuaiyu PTAL again,thanks.

**Comment by coocood** [comment]
LGTM

File: ast/functions.go [hunk 10]
```diff
@@ -329,12 +329,18 @@ func (n *FuncCallExpr) Format(w io.Writer) {
 // specialFormatArgs formats argument list for some special functions.
 func (n *FuncCallExpr) specialFormatArgs(w io.Writer) bool {
 	switch n.FnName.L {
-	case "date_add":
+	case DateAdd, DateSub, AddDate, SubDate:
```
Comment (shenli) [inline ast/functions.go]: Add some test case for this.

File: ast/functions.go [hunk 11]
```diff
@@ -329,12 +329,18 @@ func (n *FuncCallExpr) Format(w io.Writer) {
...
 		n.Args[0].Format(w)
 		fmt.Fprintf(w, ", INTERVAL ")
 		n.Args[1].Format(w)
 		fmt.Fprintf(w, " %s", n.Args[2].GetDatum().GetString())
 		return true
+	case TimestampAdd, TimestampDiff:
```
Comment (shenli) [inline ast/functions.go]: ditto

Comment (hicqu) [reply ast/functions.go]: I have add cases for `date_add` and `timestampdiff`. `date_sub` is same with `date_add`, and `timestampadd` is same with `timestampdiff`, so I think we can ignore them.

File: ast/format_test.go [hunk 12]
```diff
@@ -24,13 +24,23 @@ func (ts *testAstFormatSuite) TestAstFormat(c *C) {
...
 		{`null`, `NULL`},
 		{`"Hello, world"`, `"Hello, world"`},
 		{`'Hello, world'`, `"Hello, world"`},
 		{`'Hello, "world"'`, `"Hello, \"world\""`},
+		{`_utf8'你好'`, `"你好"`},
+		{`x'bcde'`, "x'bcde'"},
+		{`x''`, "x''"},
+		{`x'0035'`, "x'0035'"}, // Shouldn't trim leading zero.
+		{`b'00111111'`, `b'111111'`},
+		{`time'10:10:10.123'`, `timeliteral("10:10:10.123")`},
```
Comment (XuHuaiyu) [inline ast/format_test.go]: ditto

File: ast/format_test.go [hunk 13]
```diff
@@ -24,13 +24,23 @@ func (ts *testAstFormatSuite) TestAstFormat(c *C) {
 		output string
 	}{
 		// Literals.
+		{`null`, `NULL`},
+		{`true`, `TRUE`},
 		{`350`, `350`},
 		{`345.678`, `345.678`},
-		{`1e-12`, `1e-12`},
+		{`1e-12`, `0.000000000001`},
```
[the remark this answers is not in the document]
Comment (hicqu) [reply ast/format_test.go]: Hard to keep same with MySQL. After we get the float value, we have already parse the origin text into the number, so we don't know the format.

[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply ast/format_test.go]: Why not get the text in parser.y as before

[the remark this answers is not in the document]
Comment (hicqu) [reply ast/format_test.go]: There are too many places returns `ValueExpr` to `SetText` for all of them in parser. Even if we do it now, if some new grammar is added later, it's hard to force call `SetText`.
So I think convert `ValueExpr` back to text is better, though something can't be 100% same with original input.

**Comment by hicqu** [comment]
@XuHuaiyu PTAL thanks.

**Comment by XuHuaiyu** [comment]
/run-all-tests
LGTM

**Pull request merged**